iteration.py

```diff
@@ -0,0 +1,212 @@
+    )
+    url = fields.Function(fields.Char('URL'), 'get_url')
+
+    def get_url(self, name):
+        """
+        Return the url if within an active request context or return None
+        :param name: name of field
+        """
+        if has_request_context():
+            return url_for(
+                'project.iteration.render', active_id=self.id
+            )
+        return None
+
+    @classmethod
+    def __setup__(cls):
+        super(Iteration, cls).__setup__()
+        cls._order.insert(0, ('start_date', 'ASC'))
+        cls._error_messages.update({
+            'wrong_dates': 'End Date should be greater than Start Date or '
+            'Iteration period is overlaping',
+        })
+
+    @classmethod
+    def validate(cls, iterations):
+        """
+        Validate iterations
+        """
+        super(Iteration, cls).validate(iterations)
+
+        for iteration in iterations:
+            iteration.check_dates()
+
+    def serialize(self, purpose=None):
+        """
+        Serialize Iteration
+        """
+        res = {
+            'name': self.name,
+            'start_date': self.start_date.isoformat(),
+            'end_date': self.end_date.isoformat(),
+        }
+        if purpose == "full":
+            res.update({
+                'tasks': [task.serialize(purpose) for task in self.tasks]
+            })
+        return res
+
+    def check_dates(self):
+        """
+        Checks End date should not be less than Start date and,
+        Non overlapping of dates for iterations.
+        """
+        cursor = Transaction().cursor
+        table = self.__table__()
+
+        if self.start_date and self.end_date and \
+                self.start_date >= self.end_date:
+            return False
+        cursor.execute(*table.select(
+            table.id, where=(((
+                    (table.start_date <= self.start_date) &
+                    (table.end_date >= self.start_date)) |
+                ((table.start_date <= self.end_date) &
+                    (table.end_date >= self.end_date)) |
+                ((table.start_date >= self.start_date) &
+                    (table.end_date <= self.end_date))) &
+                (table.id != self.id)
+            )
+        ))
+        if cursor.fetchone():
+            return False
+        return True
+
+    @classmethod
+    @route("/iteration/", methods=["GET", "POST"])
+    @login_required
+    def render_list(cls):
+        """
+        GET: Renders list of iterations.
+        POST: Create new iteration.
+        """
+        if request.method == 'GET':
+            page = request.args.get('page', 1, int)
+            iterations = Pagination(cls, [], page, 10)
+            return jsonify(iterations.serialize())
+
+        if not current_user.has_permissions(['project.scrum_master']):
+            # Only get is allowed if user is not scrum_master.
+            abort(400)
+
+        if request.method == 'POST':
```

Comment:
use a form and validate_on_submit - takes a lof of the shit away. Also makes it easy to send errors back.